octoprint_RewriteM600/__init__.py

```diff
@@ -11,41 +12,82 @@
-			cmd = [("M117 Filament Change",),"G91","M83", "G1 Z+"+str(self._settings.get(["zDistance"]))+" E-0.8 F4500", "M82", "G90", "G1 X0 Y0"]
+			self.listening = True  # We need to listen to the first result of M114, so we can cache position
+			cmd = ["M114", ("M117 Filament Change",), "G91", "M83", "G1 E-2 F2700", "G1 Z0.05", "G1 X5 Y5", "G1 Z+" + str(self._settings.get(["zDistance"])) + " F4500", "M82", "G90", "G1 X0 Y0 F4500"]
+			self._logger.info(self.cached_position)
+			self.waiting = True
+
 		return cmd
 
-	def after_resume(self, comm_instance, phase, cmd, parameters, tags=None, *args, **kwargs):
-		if cmd and cmd == "resume":
-			if(comm_instance.pause_position.x):
-				cmd = []
-				cmd =["M83","G1 E-0.8 F4500", "G1 E0.8 F4500", "G1 E0.8 F4500", "M82", "G90", "G92 E"+str(comm_instance.pause_position.e), "M83", "G1 X"+str(comm_instance.pause_position.x)+" Y"+str(comm_instance.pause_position.y)+" Z"+str(comm_instance.pause_position.z)+" F4500"]
-				if(comm_instance.pause_position.f):
+	def detect_position(self, comm_instance, line, *args, **kwargs):
+		match = re.match("X:([0-9.]+) Y:([0-9.]+) Z:([0-9.]+) E:-?([0-9.]+) Count X:([0-9]+) Y:([0-9]+) Z:([0-9]+)", line)
+		if match is not None and self.listening:
+			self.cached_position["x"] = match.group(1)
+			self.cached_position["y"] = match.group(2)
+			self.cached_position["z"] = match.group(3)
+			self.cached_position["e"] = match.group(4)
+			self._plugin_manager.send_plugin_message(self._identifier, dict(type="popup", msg = "Saved location at X:" +
+								self.cached_position["x"] + " Y:" + self.cached_position["y"] +
+								" Z:" + self.cached_position["z"]))
+			self.listening = False  # Reset so we don't listen to the update called on print pause
+			self._logger.info(self.cached_position)
+		return line
+
+	def after_resume(self, comm_instance, script_type, script_name, *args, **kwargs):
+		self._logger.info("Received queued command: " + script_name)
+		if self.waiting and script_name == "beforePrintResumed":
+			self._logger.info("Resuming from Filament Change")
+			self.waiting = False
+			self._plugin_manager.send_plugin_message(self._identifier, dict(type = "popup", msg = "Resuming to location at X:" +
+								self.cached_position["x"] + " Y:" + self.cached_position["y"] +
+								" Z:" + self.cached_position["z"]))
+			if self.cached_position["x"] != "NOT SET":  # use our cached position
+				cmd = [
+					# We'll assume that the user manually inserted and purged the new filament, so no new extrusion
+					# is required here
+					# "M83", "G1 E-0.8 F4500", "G1 E0.8 F4500", "G1 E0.8 F4500",
+					# "M83", "G1 Z-" + str(self._settings.get(["zDistance"])) + " F4500",
+					"M82", "G90",  # Reset to absolute positioning
+					"G92 E" + str(self.cached_position["e"]),
+					# Reset our position to pre-M600 position
+					"G1 Z" + str(self.cached_position["z"]),
+					"G1 X" + str(self.cached_position["x"]) +
+					" Y" + str(self.cached_position["y"]) + " F4500"]
```

Comment:
Before this PR, the plugin used to reset to relative extruder. #11 asks to make that configurable, but as this is not the subject of this PR, I would suggest keeping the previous behaviour.

```suggestion
				cmd = [
					# We'll assume that the user manually inserted and purged the new filament, so no new extrusion
					# is required here
					# "M83", "G1 E-0.8 F4500", "G1 E0.8 F4500", "G1 E0.8 F4500",
					# "M83", "G1 Z-" + str(self._settings.get(["zDistance"])) + " F4500",
					"M82", "G90",  # Reset to absolute positioning
					"G92 E" + str(self.cached_position["e"]),
					# Reset our position to pre-M600 position
					"G1 Z" + str(self.cached_position["z"]),
					"G1 X" + str(self.cached_position["x"]) +
					" Y" + str(self.cached_position["y"]) + " F4500",
					"M83"
					]
```